Dev

daudmammaev opened the pull request:
Added implementation of writable,tests, Main examples

Commit:
001

Added implementation of writable,tests, Main examples

GrafRogi commented on .idea/.gitignore:
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
+/shelf/
```
Дополнить gitignore файлами .idea

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws IOException если возникла ошибка при записи в файл
+     */
+    @Override
+    public void writeToFile(List<?> data, String fileName) throws IOException{
```
Проверка что fileName не NULL

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -0,0 +1,46 @@
+                    if (field.isAnnotationPresent(CsvField.class)) {
+                        field.setAccessible(true);
+                        try {
+                            line.append(field.get(item)).append(",");
```
Проблемы с форматом CSV

GrafRogi commented on src/main/java/test/java/WritableTest.java:
```diff
@@ -0,0 +1,30 @@
+public class WritableTest {
+
+    @Test
+    public void testWriteToFile() throws IOException {
```
Тестов мало, они не покрывают возможные крайние случаи

GrafRogi commented on src/main/java/test/java/WritableTest.java:
```diff
@@ -0,0 +1,30 @@
+        writable.writeToFile(Arrays.asList(person1, person2), "users.csv");
+
+        // Проверка, что файл создан и содержит данные
+        assertTrue(Files.exists(Paths.get("generated.csv")));
```
![image](https://github.com/user-attachments/assets/856dd52c-5901-4d64-a17e-60073c543b7a)
Тест не работает

Commit:
002

Fixed the filename check, added new tests, fixed the format issue in csv, filled in gitignore

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -21,19 +21,32 @@ public class WritableImpl implements Writable{
     public void writeToFile(List<?> data, String fileName) throws IOException{
-        if (data == null || data.isEmpty()) {
-            throw new IllegalArgumentException("Data list cannot be null or empty");
+        if (data == null || data.isEmpty() || fileName == null) {
```
Лучше разделять это на две разные ошибки, потому что у них разная природа и последствия. И потом тяжело понять, что именно сломалось или одно или другое

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -0,0 +1,60 @@
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
+
+            // Запись заголовков
+            Field[] fields = data.get(0).getClass().getDeclaredFields();
```
Что будет если  data = NULL?

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -0,0 +1,60 @@
+                    header.append(annotation.name()).append(";");
+                }
+            }
+            writer.write(header.substring(0, header.length() - 1)); // Убираем последнюю запятую
```
Может быть ошибка, если header - будет пустой строкой. Это возможно если нет полей с аннотацией CsvField

GrafRogi commented on src/main/java/org/writer/WritableImpl.java:
```diff
@@ -0,0 +1,60 @@
+                        field.setAccessible(true);
+                        try {
+                            line.append(field.get(item)).append(";");
+                        } catch (IllegalAccessException e) {
```
Если field.get(item) = null в файле CSV тоже будет null, а должна быть пустая строка

Commit:
003

change check exception filename, added a check for null fields.Added a check if there are no fields wetted by the annotation at all

Commit:
Update WritableImpl.java

Commit:
Update WritableImpl.java
